src/symbol/collision_index.js

```diff
@@ -0,0 +1,346 @@
+        // This is a quick and dirty solution for chosing which collision circles to use (since collision circles are
+        // laid out in tile units). Ideally, I think we should generate collision circles on the fly in viewport coordinates
+        // at the time we do collision detection.
+
+        // incidenceStretch is the ratio of how much y space a label takes up on a tile while drawn perpendicular to the viewport vs
+        //  how much space it would take up if it were drawn flat on the tile
+        // Using law of sines, camera_to_anchor/sin(ground_angle) = camera_to_center/sin(incidence_angle)
+        // Incidence angle 90 -> head on, sin(incidence_angle) = 1, no stretch
+        // Incidence angle 1 -> very oblique, sin(incidence_angle) =~ 0, lots of stretch
+        // ground_angle = u_pitch + PI/2 -> sin(ground_angle) = cos(u_pitch)
+        // incidenceStretch = 1 / sin(incidenceAngle)
+
+        const incidenceStretch = pitchWithMap ? 1 : cameraToAnchorDistance / this.pitchfactor;
+        const lastSegmentTile = tileDistance.lastSegmentViewportDistance * tileToViewport;
+        return tileDistance.prevTileDistance +
+            lastSegmentTile +
+            (incidenceStretch - 1) * lastSegmentTile * Math.abs(Math.sin(lastSegmentAngle));
+    }
+
+    placeCollisionCircles(collisionCircles: Array<any>,
+                          allowOverlap: boolean,
+                          scale: number,
+                          pixelsToTileUnits: number,
+                          key: string,
+                          symbol: any,
+                          lineVertexArray: any,
+                          glyphOffsetArray: any,
+                          fontSize: number,
+                          posMatrix: mat4,
+                          labelPlaneMatrix: mat4,
+                          showCollisionCircles: boolean,
+                          pitchWithMap: boolean): Array<number> {
+        const placedCollisionCircles = [];
+
+        const projectedAnchor = this.projectAnchor(posMatrix, symbol.anchorX, symbol.anchorY);
+
+        const projectionCache = {};
+        const fontScale = fontSize / 24;
+        const lineOffsetX = symbol.lineOffsetX * fontSize;
+        const lineOffsetY = symbol.lineOffsetY * fontSize;
+
+        const tileUnitAnchorPoint = new Point(symbol.anchorX, symbol.anchorY);
+        const labelPlaneAnchorPoint = projection.project(tileUnitAnchorPoint, labelPlaneMatrix).point;
+        const firstAndLastGlyph = projection.placeFirstAndLastGlyph(
+            fontScale,
+            glyphOffsetArray,
+            lineOffsetX,
+            lineOffsetY,
+            /*flip*/ false,
+            labelPlaneAnchorPoint,
+            tileUnitAnchorPoint,
+            symbol,
+            lineVertexArray,
+            labelPlaneMatrix,
+            projectionCache,
+            /*return tile distance*/ true);
+
+        if (projectedAnchor.perspectiveRatio < .2 || !firstAndLastGlyph) {
```

Comment:
0.2 comes from just trying things out and seeing what works?

Reply:
Err... actually this is mostly left over from an earlier version of `approximateTileDistance` that had more trouble with this (overzoomed) situation. It's still true, I think, that anything with a perspective ratio of less than .2 is anchored way off screen, but it's probably not an important performance optimization so it's probably better just to remove the logic.

Reply:
I just removed the perspectiveRatio logic, and rolled the `firstAndLastGlyph` logic in with the later tile distance check. This means we do some extra work projecting collision circles for line labels way outside the viewport, but I don't think that was ever too big a deal (and if we really wanted to optimize for that case, I think I'd base it directly on the anchor position relative to the viewport rather than the perspective ratio).